resourcex/decode.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcex
+
+import (
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// Decode decodes a property map into a JSON-like structure containing only values.
+// Unknown values are decoded as nil, both in maps and arrays.
+// Secrets are collapsed into their underlying values.
+func Decode(props resource.PropertyMap) interface{} {
+	return decodeM(props)
+}
+
+// decodeV returns a mapper-compatible object map, suitable for deserialization into structures.
+func decodeM(props resource.PropertyMap) map[string]interface{} {
+	obj := make(map[string]interface{})
+	for _, k := range props.StableKeys() {
+		key := string(k)
+		obj[key] = decodeV(props[k])
+	}
+	return obj
+}
+
+// decodeV returns a mapper-compatible object map, suitable for deserialization into structures.
```

Comment:
This function is similar to the venerable `MapRepl` function, which lacks recursion into secret and output values ([code](https://github.com/pulumi/pulumi/blob/ff6404a55f0d0e9c20ef94fcce2217e343145466/sdk/go/common/resource/properties.go#L611C10-L611C25)).